config.md
# Configuration Documentation

User can change the application's configurations by modifying the user's config file (default to be `$HOME/.config/hn-tui.toml`).

**Note**: user doesn't need to specify all the options in the config file as a **default** value will be used for non-specified options.

An example config file (with some default config values) can be found in [example `hn-tui.toml`](https://github.com/aome510/hackernews-TUI/blob/main/examples/hn-tui.toml).

## Table of Contents

- [General](#general)
  - [Article Parse Command](#article-parse-command)
- [Theme](#theme)
  - [Default Theme](#default-theme)
  - [Palette](#palette)
  - [Component Style](#component-style)
- [Keymap](#keymap)
  - [Custom Keymap](#custom-keymap)
  - [Supported Keys](#supported-keys)

## General

| Option                  | Description                                                                                                           | Default                                                      |
| ----------------------- | --------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------ |
| `use_page_scrolling`    | whether to enable page-like scrolling behavior, which automatically adjusts the view based on the scrolling direction | `true`                                                       |
| `use_pacman_loading`    | whether to use a pacman loading screen or a plain loading screen                                                      | `true`                                                       |
| `url_open_command`      | the command the application uses to open an url in browser                                                            | `{ command = 'open', options = [] }`                         |
| `article_parse_command` | the command the application uses to parse an article into a readable text                                             | `{ command = 'article_md', options = ['--format', 'html'] }` |
| `client_timeout`        | the timeout (in seconds) when the application's client makes an API request                                           | `32`                                                         |

### Article Parse Command

To enable viewing an article's content in reader mode with `ArticleView`, user will need to install **additional** tools and specify the `article_parse_command` config option.

An `article_parse_command` must be a command that returns result with the following schema:

```typescript
type result_schema = {
  content: string;
  url: string;
  title: string;
  author: string | null;
  date_published: string | null;
};
```

The returned `content` **must** be a **HTML string** respresenting the article's content in reader mode.

By default, `hackernews-TUI` uses [`article_md`](https://github.com/aome510/article-md-cli) as the default command for parsing articles.

One alternative is [`mercury-parser`](https://github.com/postlight/mercury-parser#installation):

```toml
article_parse_command = { command = 'mercury-parser', options = [] }
```

## Theme

An application's theme has two components:

1. `palette`: the theme's color palette
2. `component_style`: styles for application's components

### Default theme

The default theme configurations:

```toml
[theme.palette]
background = "#f6f6ef"
foreground = "#242424"
selection_background = "#d8dad6"
selection_foreground = "#4a4c4c"
black = "#000000"
blue = "#0000aa"
cyan = "#00aaaa"
green = "#00aa00"
magenta = "#aa00aa"
red = "#aa0000"
white = "#aaaaaa"
yellow = "#aaaa00"
light_black = "#555555"
light_white = "#ffffff"
light_red = "#ff5555"
light_magenta = "#5555ff"
light_green = "#55ff55"
light_cyan = "#55ffff"
light_blue = "#5555ff"
light_yellow = "#ffff55"

[theme.component_style]
# styles for application's specific components
title_bar = { back = "#ff6600", effect = "bold" }
matched_highlight = { front = "black", back = "#ffff55"}
metadata = { front = "#828282" }
username = { effect = "bold" }
loading_bar = { front = "light yellow", back = "blue"}

# general component styles
header = { front = "black", effect = "bold" }
quote = { front = "#677280" }
italic = { effect = "italic" }
bold = { effect = "bold" }
single_code_block = { front = "black", back = "#c8c8c8"}
multiline_code_block = { front = "light black", effect = "bold" }
link = { front = "#4fbbfd" }
link_id = { front = "black", back = "#ffff55"}

# story tag styles
current_story_tag = { front = "light white" }
ask_hn = { front = "red", effect = "bold" }
tell_hn = { front = "yellow", effect = "bold" }
show_hn = { front = "blue", effect = "bold" }
launch_hn = { front = "green", effect = "bold" }
```

### Palette

A theme's color palette is based on the 4-bit ANSI terminal colors.

User can change the default color palette by changing the values of any of the following fields

- `background`
- `foreground`
- `selection_background`
- `selection_foreground`
- `black`
- `blue`
- `cyan`
- `green`
- `magenta`
- `red`
- `white`
- `yellow`
- `light_black`
- `light_blue`
- `light_cyan`
- `light_green`
- `light_magenta`
- `light_red`
- `light_white`
- `light_yellow`

Each field's value can be either a **raw hex string** representing the color (`0xf6f6ef`, `#f6f6ef`, `f6f6ef`) or a **16-bit color's name** (`black`, `dark black`, `light black`).

Specifying the 16-bit color's name will use **the terminal's default color**. For example, `background = "black"` will make the application's background be in the terminal's default black color.

### Component Style

The application defines styles for some components. For example, any links in `hackernews-TUI` has the following style by default: `link = { front = "#4fbbfd" }`.

A style has 3 **optional** fields: `front` (foreground color), `back` (background color), `effect` (additional terminal effect).

- `front` and `back` can be either a **raw hex string** representing the color (`0xf6f6ef`, `#f6f6ef`, `f6f6ef`) or a **16-bit color's name** (`black`, `dark black`, `light black`).

Specifying the 16-bit color's name will use **the theme palette's color** (as opposed to **the terminal's default color** when configuring the theme's palette color). For example, `link = { back = "blue" }` will make any links in the application use the theme palette's blue color as the background color.

- `effect` can be only **one** of the following values
  - `simple`
  - `reverse`
  - `bold`
  - `italic`
  - `strikethrough`
  - `underline`
  - `blink`

## Keymap

To modify the [default key mapping](https://github.com/aome510/hackernews-TUI#default-shortcuts), simply add new mapping entries to the corresponding keymap section. For example, to change the key shortcut for the command `next_comment` to `J` and the command `prev_comment` to `K` in the comment view, add the below lines to the config file:

```toml
[keymap.comment_view_keymap]
next_comment = "J"
prev_comment = "K"
```

It's possible for a command to have multiple keybindings. For example, to use either `j` or `J` for the `next_comment` command, add the below lines to the config file:

```toml
[keymap.comment_view_keymap]
next_comment = ["j", "J"]
```

### Custom Keymap

`custom_keymaps` is a config option used to define custom shortcuts to navigate between different story views with stories filtered by certain conditions.

`custom_keymaps` has the following schema:

```typescript
type custom_keymaps_schema = [
  {
    key: string;
    tag: "story" | "ask_hn" | "show_hn" | "job";
    by_date: bool;
    numeric_filters: {
      elapsed_days_interval: { start: number; end: number };
      points_interval: { start: number; end: number };
      num_comments_interval: { start: number; end: number };
    };
  }
];
```

An example of defining such custom keymaps can be found in the [example configuration file](https://github.com/aome510/hackernews-TUI/blob/main/examples/hn-tui.toml).

### Supported keys

List of supported keys for mapping:

- `<char>` (any single character)
- `C-<char>` (ctrl + character)
- `M-<char>` (alt + character)
- `enter`
- `tab`
- `backspace`
- `esc`
- `left`
- `right`
- `up`
- `down`
- `ins`
- `del`
- `home`
- `end`
- `page_up`
- `page_down`
- `f1`
- `f2`
- `f3`
- `f4`
- `f5`
- `f6`
- `f7`
- `f8`
- `f9`
- `f10`
- `f11`
- `F12`


hackernews_tui/src/config.rs
use anyhow::Result;
use config_parser2::*;
use cursive::theme;
use log::warn;
use once_cell::sync::OnceCell;
use serde::{de, Deserialize, Deserializer};

use super::keybindings::*;

#[derive(Debug, Deserialize, ConfigParse)]
/// Config is a struct storing the application's configurations
pub struct Config {
    pub allow_unicode: bool,
    pub page_scrolling: bool,
    pub scroll_offset: usize,
    pub url_open_command: String,
    pub article_parse_command: ArticleParseCommand,
    pub client: Client,
    pub theme: Theme,
    pub keymap: KeyMap,
}

#[derive(Debug, Deserialize, Clone, ConfigParse)]
pub struct LazyLoadingComments {
    pub num_comments_init: usize,
    pub num_comments_after: usize,
}

#[derive(Debug, Deserialize, Clone, ConfigParse)]
pub struct ArticleParseCommand {
    pub command: String,
    pub options: Vec<String>,
}

#[derive(Debug, Deserialize, ConfigParse)]
pub struct StoryLimit {
    pub front_page: usize,
    pub story: usize,
    pub ask_hn: usize,
    pub show_hn: usize,
    pub job: usize,
    pub search: usize,
}

#[derive(Debug, Deserialize, ConfigParse)]
pub struct Client {
    pub story_limit: StoryLimit,
    pub lazy_loading_comments: LazyLoadingComments,
    pub client_timeout: u64,
}

#[derive(Debug, Clone)]
pub struct Color {
    pub color: theme::Color,
}

impl Color {
    fn parse(s: &str) -> Option<Self> {
        theme::Color::parse(s).map(|color| Color { color })
    }
}

impl StoryLimit {
    pub fn get_story_limit_by_tag(&self, tag: &str) -> usize {
        match tag {
            "front_page" => self.front_page,
            "story" => self.story,
            "job" => self.job,
            "ask_hn" => self.ask_hn,
            "show_hn" => self.show_hn,
            _ => panic!("unknown tag: {}", tag),
        }
    }
}

impl<'de> de::Deserialize<'de> for Color {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        match Color::parse(&s) {
            None => Err(de::Error::custom(format!("failed to parse color: {}", s))),
            Some(color) => Ok(color),
        }
    }
}

config_parser_impl!(Color);

#[derive(Debug, Deserialize, Clone, ConfigParse)]
pub struct Theme {
    // cursive's palette colors
    pub background: Color,
    pub view: Color,
    pub shadow: Color,
    pub primary: Color,
    pub secondary: Color,
    pub tertiary: Color,
    pub title_primary: Color,
    pub title_secondary: Color,
    pub highlight: Color,
    pub highlight_inactive: Color,
    pub highlight_text: Color,

    // additional custom colors
    pub link_text: Color,
    pub link_id_bg: Color,
    pub search_highlight_bg: Color,
    pub status_bar_bg: Color,
    pub code_block_bg: Color,
}

impl Theme {
    pub fn update_theme(&self, theme: &mut theme::Theme) {
        theme.palette.set_color("background", self.background.color);
        theme.palette.set_color("view", self.view.color);
        theme.palette.set_color("shadow", self.shadow.color);
        theme.palette.set_color("primary", self.primary.color);
        theme.palette.set_color("secondary", self.secondary.color);
        theme.palette.set_color("tertiary", self.tertiary.color);
        theme
            .palette
            .set_color("title_primary", self.title_primary.color);
        theme
            .palette
            .set_color("title_secondary", self.title_secondary.color);
        theme.palette.set_color("highlight", self.highlight.color);
        theme
            .palette
            .set_color("highlight_inactive", self.highlight_inactive.color);
        theme
            .palette
            .set_color("highlight_text", self.highlight_text.color);
    }
}

impl Config {
    // parse config struct from a file
    pub fn from_config_file(file_path: &str) -> Result<Self> {
        match std::fs::read_to_string(file_path) {
            // if cannot open the file, use the default configurations
            Err(err) => {
                warn!(
                    "failed to open {}: {:#?}\n...Use the default configurations instead",
                    file_path, err
                );
                Ok(Self::default())
            }
            Ok(config_str) => {
                let value = toml::from_str::<toml::Value>(&config_str)?;
                let mut config = Self::default();
                config.parse(value)?;
                Ok(config)
            }
        }
    }
}

impl Default for Theme {
    fn default() -> Self {
        Theme {
            background: Color::parse("#f6f6ef").unwrap(),
            shadow: Color::parse("#000000").unwrap(),
            view: Color::parse("#f6f6ef").unwrap(),
            primary: Color::parse("#4a4a48").unwrap(),
            secondary: Color::parse("#a5a5a5").unwrap(),
            tertiary: Color::parse("#ffffff").unwrap(),
            title_primary: Color::parse("#000000").unwrap(),
            title_secondary: Color::parse("#ffff00").unwrap(),
            highlight: Color::parse("#6c6c6c").unwrap(),
            highlight_inactive: Color::parse("#0000ff").unwrap(),
            highlight_text: Color::parse("#c3bbbb").unwrap(),

            link_text: Color::parse("#4fbbfd").unwrap(),
            link_id_bg: Color::parse("#ffff00").unwrap(),
            search_highlight_bg: Color::parse("#ffff00").unwrap(),
            status_bar_bg: Color::parse("#ff6600").unwrap(),
            code_block_bg: Color::parse("#c8c8c8").unwrap(),
        }
    }
}

impl Default for Config {
    fn default() -> Self {
        Config {
            allow_unicode: false,
            page_scrolling: true,
            scroll_offset: 3,
            url_open_command: "xdg-open".to_string(),
            article_parse_command: ArticleParseCommand {
                command: "mercury-parser".to_string(),
                options: vec!["--format".to_string(), "markdown".to_string()],
            },
            client: Client {
                lazy_loading_comments: LazyLoadingComments {
                    num_comments_init: 5,
                    num_comments_after: 10,
                },
                story_limit: StoryLimit {
                    search: 10,
                    front_page: 20,
                    story: 20,
                    ask_hn: 15,
                    show_hn: 15,
                    job: 15,
                },
                client_timeout: 32,
            },
            theme: Theme::default(),
            keymap: KeyMap::default(),
        }
    }
}

static CONFIG: OnceCell<Config> = OnceCell::new();

pub fn init_config(config: Config) {
    CONFIG.set(config).unwrap_or_else(|_| {
        panic!("failed to set up the application's configurations");
    });
}

pub fn get_config() -> &'static Config {
    &CONFIG.get().unwrap()
}

pub fn get_config_theme() -> &'static Theme {
    &get_config().theme
}

pub fn get_custom_keymap() -> &'static CustomKeyMap {
    &get_config().keymap.custom_keymap
}

pub fn get_global_keymap() -> &'static GlobalKeyMap {
    &get_config().keymap.global_keymap
}

pub fn get_story_view_keymap() -> &'static StoryViewKeyMap {
    &get_config().keymap.story_view_keymap
}

pub fn get_search_view_keymap() -> &'static SearchViewKeyMap {
    &get_config().keymap.search_view_keymap
}

pub fn get_comment_view_keymap() -> &'static CommentViewKeyMap {
    &get_config().keymap.comment_view_keymap
}

pub fn get_article_view_keymap() -> &'static ArticleViewKeyMap {
    &get_config().keymap.article_view_keymap
}
